docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,1186 @@
+        value: foo
+        ```
+
+    1. To point to some external data for a given storage backend:
+
+        ```yaml
+        # option 2: register something which already exist as external TypeInstance - based on `backend.additionalParameters`
+        backend:
+          additionalParameters:
+            name: release-name
+            namespace: release-namespace
+        ```
+
+        However, the `additionalParameters` are backend-specific properties, which means Content Developer need to explicitly specify the backend as described later.
+
+    1. To save a specific value with additional parameters:
+    
+        For example, for an implementation of Kubernetes secrets storage backend, which actually creates and updates these secrets during TypeInstance creation:
+
+        ```yaml
+        # option 3: save a specific value on an external backend with some additional parameters
+        value: foo
+        backend:
+          additionalParameters:
+            key: bar
+            value: baz
+        ```
+
+        The storage backend has to have `additionalParametersSchema` specified, as well as the `acceptValue` property set to `true`.
+
+    In that way, someday we will be able to extend such approach with additional properties:
+    
+    ```yaml    
+    instrumentation: # someday - if we want to unify the approach
+      health:
+        endpoint: foo.bar/healthz
+      # (...)
+    ```
+
+    Such `instrumentation` data would be also stored in the same storage backend as the `value`. If Content Developer wants to store it somewhere else, then an additional Argo artifact to produce is needed.
+
+1. Then, Content Developer specifies the Argo workflow artifact as output TypeInstance with familiar syntax:
+
+    ```yaml
+    # default - static
+    capact-outputTypeInstances:
+      - name: mattermost-config
+        from: additional
+        # no backend definition -> use default storage backend
+
+    # option 2 - specific backend (defined in `Implementation.spec.requires` property)
+    capact-outputTypeInstances:
+      - name: helm-release
+        from: helm-release
+        backend: helm-storage # new property -> alias defined in the `requires` section
+    ```
+
+1. The automatically injected TypeInstance upload step, receives the following payload:
+
+    ```yaml
+    typeInstances:
+    - alias: helm-release
+      attributes: []
+      createdBy: default/act-l49vh-30c7a078-6a77-475c-94dd-7466f56447ce
+      typeRef:
+        path: cap.type.helm.chart.release
+        revision: 0.1.0
+      value: null 
+      backend:
+        # Fields which are a part of TypeInstance
+        id: 3ef2e4ac-9070-4093-a3ce-142139fd4a16 # helm-release backend - resolved UUID based on the injected TypeInstance
+
+        # Fields which are a part of TypeInstanceResourceVersion (can be changed later via TypeInstance Update):
+        additionalParameters:
+          name: release-name
+          namespace: release-namespace
```

Comment:
Not sure what you mean 🤔 

This is exactly what will be inside the Argo artifact, and this will be passed as the payload to GraphQL API. Content Dev need to produce such TypeInstance to point to a proper helm release.

And yes, the. `value` is null in this context. Some of the storage backend can require just `value`, others just `additionalParameters` and others both.

```yaml
value: null
backend:
     additionalParameters:
            name: release-name
            namespace: release-namespace
```

Reply:
>This is exactly what will be inside the Argo artifact, 

exactly :D

but in your proposal I see:
> 1. The automatically injected TypeInstance upload step, receives the following payload:

so this won't be in payload at all, as we will have only information about artifact name (which also matches the TypeInstance alias)

so as a wrote:
> AFAIK this data will be in artifacts, we could render it here, but this would require changes also in manifest.

probably you don't want to render that, correct? so in payload we will not have anything about the `additionalParameters` property

Reply:
It will be payload indeed - for GraphQL API. To dicuss f2f, maybe I don't understand the problem 🤔